enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,321 @@
+
+1. Copying AMIs from us-east-1 to the region where the user is installing.
+Currently the installer only copies and encrypts the AMI published by RHCOS in the same region for low latency. But the installer should be able to support copying and encrypting the AMI published by RHCOS for us-east-1 region to the region user is installing the cluster. This will cause the time to copy to the AMI to that region, but it should enable users to user these new regions.
+
+2. Allowing users to provide the AMI in the install-config for regions that cannot import AMIs from other regions like US Gov Cloud.
+This is already supported by [install-config][install-config-aws-ami]. The users will have to use the user-provided-installation workflow to get the AMI imported to the region where the cluster will be installed.
+The installer will use the AMI provided and copy-encrypt to an AMI for the cluster, keeping the behavior consistent with other workflows.
+
+To support custom endpoints for AWS APIs, the install-config would allow the users to provide a list of endpoints for the services. The endpoints will be used by the installer to call AWS APIs to perform validation, create the MachineSet objects, and will be used to configure the terraform AWS provider. When the custom endpoints will be set, the installer will validate that the endpoints are reachable. The installer requires AWS API endpoints for EC2, IAM, Route53, ELB, ELBv2, STS, but the installer will allow the user to provide API endpoints for a subset of the endpoints under the assumption that default API endpoints are usable.
+
+When custom endpoints are provide for AWS APIs, the cluster operators also need the discover the information, therefore, the installer will set the AWS API endpoints in the `spec` of the `infrastructures.config.openshift.io` `cluster` object.
+
+### User Stories
+
+#### Using private VPC endpoints
+
+#### Installing to US Gov Cloud
+
+#### Using FIPS endpoints for US Gov Cloud
+
+#### Installing to custom region
+
+#### Installing to new regions without native RHCOS support in the region
+
+### Implementation Details/Notes/Constraints
+
+#### Install Config
+
+```go
+// CustomEndpoint store the configuration of a custom url to
+// override existing defaults of AWS Services.
+// Currently Supports - EC2, IAM, ELB, S3 and Route53.
+type CustomEndpoint struct {
+    Service string `json:"service"`
+
+    // This must be a HTTPS URL
+    URL     string `json:"url"`
+}
+```
+
+```go
+type Platform struct {
+    ...
+    // APIEndpoints list contains custom endpoints which will override default
+    // service endpoint of AWS Services.
+    // There must be only one CustomEndpoint for a service.
+    // +optional
+    APIEndpoints []CustomEndpoint `json:"customRegionOverride,omitempty"`
+    ...
+}
+```
+
+#### Destroy Cluster
+
+The `metadata.json` needs to store the custom endpoints from the install-config.yaml so that the users can delete the clusters without any previous state.
```

Comment:
Is this reused by hive?

Reply:
hive launches a deprovision in a way that doesn't use metadata.json https://github.com/openshift/hive/blob/f12a23ba3ee502ed8e71761e1d1af3c080e3adad/pkg/install/generate.go#L435-L451

but hive would need to save these custom endpoints so that a deprovision could succeed.

CC @dgoodwin